crates/primitives/src/credential.rs

```diff
@@ -162,6 +164,96 @@ impl Credential {
+        self.associated_data_hash = Self::hash_bytes_to_field_element(data)?;
+        Ok(self)
+    }
+
+    /// Hashes arbitrary bytes to a field element using Poseidon2 sponge construction.
+    ///
+    /// This uses a sponge construction to support **arbitrary length** input:
+    /// 1. Split input into 31-byte chunks (each fits safely in a field element)
+    /// 2. Absorb chunks in batches of up to 8 elements (rate) into the state
+    /// 3. Apply Poseidon2 t16 permutation after each batch
+    /// 4. Apply padding + domain separation (constant tag and length) and squeeze
+    ///
+    /// The state is divided into:
+    /// - Rate portion (indices 0-7): where data is absorbed via addition
+    /// - Capacity portion (indices 8-15): provides security, not directly modified by input
+    ///
+    /// # Arguments
+    /// * `data` - Arbitrary bytes to hash (any length).
+    ///
+    /// # Errors
+    /// Will error if the data is empty.
+    pub fn hash_bytes_to_field_element(data: &[u8]) -> Result<FieldElement, PrimitiveError> {
+        /// Number of bytes per chunk. 31 bytes = 248 bits, which fits safely in
+        /// the BN254 scalar field (< 254 bits).
+        const CHUNK_SIZE: usize = 31;
+        /// Rate: number of field elements absorbed per permutation.
+        /// Using 8 leaves 8 elements as capacity for security.
+        const RATE: usize = 8;
+
+        // Domain separation tag to avoid collisions with other Poseidon2 usages.
+        const DS_TAG: &[u8] = b"ASSOCIATED_DATA_HASH_V1";
+
+        if data.is_empty() {
+            return Err(PrimitiveError::InvalidInput {
+                attribute: "associated_data".to_string(),
+                reason: "data cannot be empty".to_string(),
+            });
+        }
+
+        let poseidon2: Poseidon2<Fq, 16, 5> = Poseidon2::new(&POSEIDON2_BN254_T16_PARAMS);
+
+        // Initialize state with zeros
+        let mut state: [Fq; 16] = [Fq::ZERO; 16];
+
+        // Apply domain separation in the capacity portion.
+        state[RATE] += *FieldElement::from_be_bytes_mod_order(DS_TAG);
+
+        // Convert bytes to field elements
+        let field_elements: Vec<Fq> = data
+            .chunks(CHUNK_SIZE)
+            .map(|chunk| *FieldElement::from_be_bytes_mod_order(chunk))
+            .collect();
+
+        // Absorb field elements in batches of RATE
+        for batch in field_elements.chunks(RATE) {
+            // Add batch elements to the rate portion of the state
+            for (i, &elem) in batch.iter().enumerate() {
+                state[i] += elem;
+            }
+            // Apply permutation after each batch
+            poseidon2.permutation_in_place(&mut state);
+        }
+
+        // Padding marks the end of data absorption; prefix-freedom is ensured by the combination
+        // of this padding and the length encoding below, which prevents collisions between inputs
+        // where one is a prefix of another.
+        state[0] += Fq::ONE;
```

Comment:
This method allows for length extension attacks, since field elements that are not exact multiples of the rate get implicitly padded with zeros above. This means that something like [123]  and [123,0,0,0,0,0,0,0] (in FE) would produce the same hash, which is not wanted.
My suggestion would be to implement the SAFE API for this: https://eprint.iacr.org/2023/522, this handles the lengths and padding as well.